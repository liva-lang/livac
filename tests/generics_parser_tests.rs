use insta::assert_snapshot;
use livac::{lexer::tokenize, parser::parse};
use std::fs;

/// Test helper for generic parser tests
fn test_generic_ok(test_name: &str) {
    let source = fs::read_to_string(format!("tests/parser/generics/ok_{}.liva", test_name))
        .unwrap_or_else(|_| panic!("Failed to read test file: ok_{}.liva", test_name));

    let tokens = tokenize(&source).unwrap();
    let ast = parse(tokens, &source).unwrap();

    // Convert AST to JSON for snapshot
    let json = serde_json::to_string_pretty(&ast).unwrap();
    assert_snapshot!(format!("generics_ok_{}.ast", test_name), json);
}

#[test]
fn test_generic_function_simple() {
    test_generic_ok("generic_function_simple");
}

#[test]
fn test_generic_function_multiple() {
    test_generic_ok("generic_function_multiple");
}

#[test]
fn test_generic_function_constraint() {
    test_generic_ok("generic_function_constraint");
}

#[test]
fn test_generic_function_multiple_constraints() {
    test_generic_ok("generic_function_multiple_constraints");
}

#[test]
fn test_generic_class_simple() {
    test_generic_ok("generic_class_simple");
}

#[test]
fn test_generic_class_multiple() {
    test_generic_ok("generic_class_multiple");
}

#[test]
fn test_generic_class_with_constraint() {
    test_generic_ok("generic_class_with_constraint");
}

#[test]
fn test_generic_method() {
    test_generic_ok("generic_method");
}

#[test]
fn test_identity_oneliner() {
    test_generic_ok("identity_oneliner");
}

#[test]
fn test_generic_type_arguments() {
    test_generic_ok("generic_type_arguments");
}

#[test]
fn test_nested_generics() {
    test_generic_ok("nested_generics");
}
