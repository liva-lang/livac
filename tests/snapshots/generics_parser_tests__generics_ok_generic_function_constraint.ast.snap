---
source: tests/generics_parser_tests.rs
expression: json
---
{
  "items": [
    {
      "Function": {
        "name": "max",
        "type_params": [
          {
            "name": "T",
            "constraint": "Comparable"
          }
        ],
        "params": [
          {
            "name": "a",
            "type_ref": {
              "Simple": "T"
            },
            "default": null
          },
          {
            "name": "b",
            "type_ref": {
              "Simple": "T"
            },
            "default": null
          }
        ],
        "return_type": {
          "Simple": "T"
        },
        "body": {
          "stmts": [
            {
              "If": {
                "condition": {
                  "Binary": {
                    "op": "Gt",
                    "left": {
                      "MethodCall": {
                        "object": {
                          "Identifier": "a"
                        },
                        "method": "compareTo",
                        "args": [
                          {
                            "Identifier": "b"
                          }
                        ],
                        "adapter": "seq",
                        "adapter_options": {}
                      }
                    },
                    "right": {
                      "Literal": {
                        "Int": 0
                      }
                    }
                  }
                },
                "then_branch": {
                  "Block": {
                    "stmts": [
                      {
                        "Return": {
                          "expr": {
                            "Identifier": "a"
                          }
                        }
                      }
                    ]
                  }
                },
                "else_branch": null
              }
            },
            {
              "Return": {
                "expr": {
                  "Identifier": "b"
                }
              }
            }
          ]
        },
        "expr_body": null,
        "is_async_inferred": false,
        "contains_fail": false
      }
    }
  ]
}
